packages/loopring_v3/contracts/amm/libamm/AmmJoinRequest.sol

```diff
@@ -0,0 +1,103 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2017 Loopring Technology Limited.
+pragma solidity ^0.7.0;
+pragma experimental ABIEncoderV2;
+
+import "../../lib/EIP712.sol";
+import "../../lib/ERC20SafeTransfer.sol";
+import "../../lib/MathUint.sol";
+import "../../lib/MathUint96.sol";
+import "../../thirdparty/SafeCast.sol";
+import "./AmmData.sol";
+import "./AmmUtil.sol";
+
+
+/// @title AmmJoinRequest
+library AmmJoinRequest
+{
+    using ERC20SafeTransfer for address;
+    using MathUint          for uint;
+    using MathUint96        for uint96;
+    using SafeCast          for uint;
+
+    function depositToPool(
+        AmmData.State storage S,
+        uint96[]     calldata amounts
+        )
+        public
+    {
+        uint size = S.tokens.length;
+        require(amounts.length == size + 1, "INVALID_DATA");
+
+        // Question(brecht): I don't understand this part, may be reasonable inside
+        // the withdrawlFromPool function, but I'm not sure it's necessary at all.
+        if (S.isExiting[msg.sender]) {
+            // This could suddenly change the amount of liquidity tokens available, which
+            // could change how the operator needs to process the exit.
+            require(amounts[0] == 0, "CANNOT_DEPOSIT_LIQUIDITY_TOKENS_WHILE_EXITING");
+        }
+
+        S.lockedUntil[msg.sender] = 0;
```

Comment:
I think users having the option to immediately lock the tokens again can also invalidate blocks:
- User unlocks funds, but keeps funds in the pool contract
- User does an onchain exit which the operator needs to process
- The funds are unlocked so the operator needs to process the exit as if no liquidity tokens are available. The operator start building the block that processes the exit
- The user now locks up all funds again, so the liquidity tokens are available again and need to be used in the exit
- The operator needs to re-do the exit and throw away any blocks/txs that depend on the wrong exit.

So I think any change from locked -> unlocked and unlocked -> locked needs to be done with a decent delay.

Reply:
Another way this could be solved is that for onchain exits `poolAmountIn` is locked up in yet another place, which can only be used by onchain exits (cannot even be used by exits that are approved with a signature) and need to either be transferred from the unlocked balance or be newly deposit directly to it. When processing the onchain exit it's either burned or moved to the unlocked funds (similar to how it previously worked with the queue).

This would partly make things easier onchain for all the locking/unlocking headaches, but we'd need to track yet another set of balances (though this should be farily easy because only used for the onchain exits, but need to be made available for withdrawing when the pool is shutdown etc...).

Reply:
Please proceed with a fix of your preference.